lib/src/main/java/io/ably/lib/transport/Hosts.java

```diff
@@ -40,39 +40,38 @@
+            }
+            if (options.port != 0 || options.tlsPort != 0) {
+                throw AblyException.fromErrorInfo(new ErrorInfo("fallbackHostsUseDefault cannot be set when port or tlsPort are set", 40000, 400));
+            }
+            tempFallbackHosts = Defaults.HOST_FALLBACKS;
+        }
+
+        boolean isProduction = options.environment == null || options.environment.isEmpty() || "production".equalsIgnoreCase(options.environment);
+
+        if (!hasCustomPrimaryHost && tempFallbackHosts == null && options.port == 0 && options.tlsPort == 0) {
+            tempFallbackHosts = isProduction ? Defaults.HOST_FALLBACKS : Defaults.getEnvironmentFallbackHosts(options.environment);
+        }
+
+        if (hasCustomPrimaryHost) {
             setPrimaryHost(primaryHost);
             if (options.environment != null) {
                 /* TO3k2: It is never valid to provide both a restHost and environment value
                  * TO3k3: It is never valid to provide both a realtimeHost and environment value */
                 throw AblyException.fromErrorInfo(new ErrorInfo("cannot set both restHost/realtimeHost and environment options", 40000, 400));
             }
-        } else if (options.environment != null && !options.environment.equalsIgnoreCase("production")) {
-            /* RSC11: If ClientOptions.environment is set and is not
-             * "production", then the primary hostname is set to the default
-             * hostname with the environment setting used as a prefix.
-             * Note that this does not happen if there is an explicit setting
-             * of ClientOptions.restHost or ClientOptions.realtimeHost (as
-             * appropriate). The spec is not clear on which one should take
-             * precedence. */
-            setPrimaryHost(options.environment + "-" + defaultHost);
         } else {
-            setPrimaryHost(defaultHost);
-        }
-        fallbackHostsUseDefault = options.fallbackHostsUseDefault;
-        if (options.fallbackHosts == null) {
-            fallbackHosts = Arrays.copyOf(Defaults.HOST_FALLBACKS, Defaults.HOST_FALLBACKS.length);
-            fallbackHostsIsDefault = true;
-        } else {
-            /* RSC15a: use ClientOptions#fallbackHosts if set */
-            fallbackHosts = Arrays.copyOf(options.fallbackHosts, options.fallbackHosts.length);
-            fallbackHostsIsDefault = false;
-            if (options.fallbackHostsUseDefault) {
-                /* TO3k7: It is never valid to configure fallbackHost and set
-                 * fallbackHostsUseDefault to true */
-                throw AblyException.fromErrorInfo(new ErrorInfo("cannot set both fallbackHosts and fallbackHostsUseDefault options", 40000, 400));
-            }
+            setPrimaryHost(isProduction ? defaultHost : options.environment + "-" + defaultHost);
         }
+
+        fallbackHostsIsDefault = Arrays.equals(Defaults.HOST_FALLBACKS, tempFallbackHosts);
```

Comment:
Does this do deepEquals 🤔?

Reply:
Yes, it will check the number of elements as well as if all of them are equal